snap/gadget.go

```diff
@@ -234,6 +256,231 @@ func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
+	// a tmp file to find out the default sector size of the device the tmp
+	// file is created on
+	return nil
+}
+
+func validateStructureType(s string, vol *GadgetVolume) error {
+	// Type can be one of:
+	// - "mbr" (backwards compatible)
+	// - "bare"
+	// - [0-9A-Z]{2} - MBR type
+	// - GPT UUID
+	// - hybrid ID
+	//
+	// Hybrid ID is 2 hex digits of MBR type, followed by 36 GUUID
+	// example: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+
+	schema := vol.Schema
+	if schema == "" {
+		schema = "gpt"
+	}
+
+	if s == "" {
+		return errors.New(`type is not specified`)
+	}
+
+	if s == "bare" {
+		// unknonwn blob
+		return nil
+	}
+
+	if s == "mbr" {
+		// backward compatibility for type: mbr
+		return nil
+	}
+
+	var isGPT, isMBR bool
+
+	idx := strings.IndexRune(s, ',')
+	if idx == -1 {
+		// just ID
+		switch {
+		case validTypeID.MatchString(s):
+			isMBR = true
+		case validGUUID.MatchString(s):
+			isGPT = true
+		default:
+			return fmt.Errorf("invalid format")
+		}
+	} else {
+		// hybrid ID
+		code := s[:idx]
+		guid := s[idx+1:]
+		if len(code) != 2 || len(guid) != 36 || !validTypeID.MatchString(code) || !validGUUID.MatchString(guid) {
+			return fmt.Errorf("invalid format of hybrid type")
+		}
+	}
+
+	if schema != "gpt" && isGPT {
+		// type: <uuid> is only valid for GPT volumes
+		return fmt.Errorf("GUID structure type with non-GPT schema %q", vol.Schema)
+	}
+	if schema != "mbr" && isMBR {
+		return fmt.Errorf("MBR structure type with non-MBR schema %q", vol.Schema)
+	}
+
+	return nil
+}
+
+func validateRole(vs *VolumeStructure, vol *GadgetVolume) error {
+	if vs.Type == "bare" {
+		if vs.Role != "" && vs.Role != "mbr" {
+			return fmt.Errorf("conflicting type: %q", vs.Type)
+		}
+	}
+	vsRole := vs.Role
+	if vs.Type == "mbr" {
+		// backward compatibility
+		vsRole = "mbr"
+	}
+
+	switch vsRole {
+	case "system-data":
+		if vs.Label != "" && vs.Label != "writable" {
+			return fmt.Errorf(`role of this kind must have an implicit label or "writable", not %q`, vs.Label)
+		}
+	case "mbr":
+		if vs.Size > 446 {
+			return errors.New("mbr structures cannot be larger than 446 bytes")
+		}
+		if vs.Offset != 0 {
+			return errors.New("mbr structure must start at offset 0")
+		}
+		if vs.ID != "" {
+			return errors.New("mbr structure must not specify partition ID")
+		}
+		if vs.Filesystem != "" && vs.Filesystem != "none" {
+			return errors.New("mbr structures must not specify a file system")
+		}
+	case "system-boot", "":
+		// noop
+	default:
+		return fmt.Errorf("unsupported role")
+	}
+	return nil
+}
+
+func validateBareContent(vc *VolumeContent) error {
+	if vc.Source != "" || vc.Target != "" {
+		return fmt.Errorf("cannot use non-image content for bare file system")
+	}
+	if vc.Image == "" {
+		return fmt.Errorf("missing image file name")
+	}
+	return nil
+}
+
+func validateFilesystemContent(vc *VolumeContent) error {
+	if vc.Image != "" || vc.Offset != 0 || vc.OffsetWrite != (GadgetRelativeOffset{}) || vc.Size != 0 {
```

Comment:
Perhaps add an IsZero method to GadgetRelativeOffset?

Reply:
There will be some more code added with actual gadget data write support. I'll add the helper there if the check comes up more than once.